blog/2020/05/invalidations.md

```diff
@@ -0,0 +1,518 @@
+
+If you add fourth and fifth methods,
+
+```
+f(::AbstractArray) = 4
+f(::Missing) = 5
+```
+
+then Julia produces
+
+```
+julia> @code_typed applyf(c)
+CodeInfo(
+1 ─ %1 = Base.arrayref(true, container, 1)::Any
+│   %2 = Main.f(%1)::Any
+│   %3 = Base.arrayref(true, container, 2)::Any
+│   %4 = Main.f(%3)::Any
+│   %5 = (%2 + %4)::Any
+└──      return %5
+) => Any
+```
+
+There are now so many possibilities that Julia just gives up and
+uses "runtime dispatch" to decide what method of `f` to call.
+It doesn't even try to enforce the fact that `f` returns and `Int`,
+in part because determining such facts takes time (adding to compiler latency)
+and because functions with many methods typically tend to return multiple types
+anyway.
+
+Compiling each of these new implementations takes JIT-time.
+If Julia knew in advance that you'd arrive at this place, it would never have bothered to produce that first, heavily-optimized version of `applyf`.
+But the performance benefits of such optimizations are so large that, when applicable, they are well worth it.
+For example, if you start a fresh Julia session and just define the `f(::Int)`
+and `f(::Bool)` methods, then
+
+```
+julia> using BenchmarkTools
+
+julia> @btime applyf($c)
+  4.659 ns (0 allocations: 0 bytes)
+3
+
+julia> f(::String) = 3
+f (generic function with 3 methods)
+
+julia> f(::AbstractArray) = 4
+f (generic function with 4 methods)
+
+julia> f(::Missing) = 5
+f (generic function with 5 methods)
+
+julia> @btime applyf($c)
+  33.537 ns (0 allocations: 0 bytes)
+3
+```
+It's almost a tenfold difference.
+If `applyf` is performance-critical, you'll be very happy that Julia tries to give you the best version it can, given the available information.
+But this leaves the door open to invalidation, which means recompilation the next time you use `applyf`.
+If method invalidation happens often, this might contribute to making Julia "feel" sluggish.
+
+# How common is method invalidation?
+
+Unfortunately, method invalidation is pretty common.
+First, let's get some baseline statistics.
+Using the [MethodAnalysis] package (which is at a very early stage of development
+at the time of this writing), you can find out that a fresh Julia session
+(albeit one that has loaded the MethodAnalysis package and used it to perform some analysis) has almost 50,000 `MethodInstance`s tucked away in its cache.
+These are mostly for `Base` and the standard libraries.
+
+Using some not-yet merged work in both Julia itself and [SnoopCompile], we can count the number of invalidations when we load various packages into a fresh Julia session:
+
+| Package | # of unique invalidations |
+|:------- | ------------------:|
+| Example | 0 |
+| Revise | 27 |
+| FixedPointNumbers | 429 |
+| SIMD | 2799 |
+| StaticArrays | 2852 |
+| Optim | 3171 |
+| Images | 3638 |
+| Flux | 3697 |
+| Plots | 4002 |
+| DataFrames | 4048 |
+| JuMP | 4666 |
+| Makie | 6118 |
+| DifferentialEquations | 6777 |
```

Comment:
oh wow. Can I see this list? I'm a bad bad man.

Reply:
It's not future-proof given the reporting will likely change, but you can just throw this action into a PR for the report https://github.com/ianshmean/Plots.jl/pull/1/files

Reply:
You're not bad, it's mostly a consequence of having lots of dependencies that extend low-level functionality. (Packages that are essentially totally isolated do not cause invalidations, it's when you start extending functionality that this becomes an issue.) It's not very hard to generate this yourself if you can build Julia, but to save you the trouble I've posted it [here](https://gist.github.com/timholy/88fe5376f0d1460b9b9d9bd253f51b99). About 50% of the nominal invalidations come from just 7 methods (out of 90 total that cause invalidations). And the packages for these 7 are:
```
 StaticArrays
 StaticArrays
 Intervals
 Unitful
 ProgressLogging
 SymbolicUtils
 Intervals
```

I should probably develop printing options that, e.g., allow you to cull the ones with few children, but for now this is the state-of-the-art.

Once Julia stops considering ambiguous type-intersections as invalidating, your total will go down a lot (estimated to fix about two-thirds of "your" total count).

Reply:
StaticArrays seems to be showing up in a lot of these. If that's in the sysimage, I wonder how many downstream packages get a faster compile time (+ the ambiguous fix).

Reply:
> If that's in the sysimage

You can just put it in your sysimage and try (PackageCompiler.jl)